File: src/math/CombinationSum_39.java
package math;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static tree.StringUtils.printDim;

public class CombinationSum_39 {

    public static void main(String[] args) {
        System.out.println(printDim(new CombinationSum_39().combinationSum(new int[]{2, 3, 6, 7}, 7)));
        System.out.println(printDim(new CombinationSum_39().combinationSum(new int[]{2, 3, 5}, 8)));
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();
        backtrack(candidates, new ArrayList<>(), result, target, 0);
        return result;
    }

    private void backtrack(int[] candidates, List<Integer> temp, List<List<Integer>> result, int remain, int start) {
        if (remain < 0) {
            return;
        } else if (remain == 0) {
            result.add(new ArrayList<>(temp));
        } else {
            for (int i = start; i < candidates.length; i++) {
                temp.add(candidates[i]);
                backtrack(candidates, temp, result, remain - candidates[i], i);
                temp.remove(temp.size() - 1);
            }
        }
    }
}

File: src/subsets/combination_sum_39.py
class Solution(object):
    def combinationSum(self, candidates, target):
        results = []
        self.combinationSumRec(candidates, results, [], 0, target)
        return results

    def combinationSumRec(self, candidates, results, comb, start, remain):
        if remain < 0:
            return
        elif remain == 0:
            results.append(list(comb))
        else:
            for i in range(start, len(candidates)):
                comb.append(candidates[i])
                self.combinationSumRec(candidates, results, comb, i, remain - candidates[i])
                comb.pop()


print(Solution().combinationSum(candidates=[2, 3, 6, 7], target=7))
print(Solution().combinationSum(candidates=[2, 3, 5], target=8))
print(Solution().combinationSum(candidates=[2], target=1))
